Added m_joinpartspam: chmode to ban joinpart spammers.

genius3000 opened the pull request:
This adds channel mode +U with a 'cycles:sec' parameter. A cycle is counted as a successful join to the channel. Upon reaching the limit, the next attempted join (prior to the time reset) results in a WildHost ban and denying the join.

I'm using a mask of 'user@host' instead of 'User*' for tracking so a user is still caught after a quit, connect, and join. Not using the nickname in the mask so it's not so easily bypassed.

GLolol commented:
Cmode +U conflicts with #64 :(

Techman- commented:
Network ops still have the ability to pick and choose what modules to use :)

Configurable modechar would be good.

genius3000 commented:
I just had a look through helpop and went with an available letter. Could try find another one.
It is easy to change, just one spot in the code (other than comments).

genius3000 commented:
Updated this with a few code style fixes and changed to chmode V. I know that could conflict with customprefix->Half voice, but I've never seen that used.

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,226 @@
+		for (std::map<std::string, time_t>::iterator i = resets.begin(); i != resets.end(); i++)
+		{
+			if (ServerInstance->Time() > resets[i->first])
+				this->clear(i->first);
```
This call invalidates the iterator, if the entry needs to be removed ++ the iterator before calling the function

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,231 @@
+	};
+	std::map<std::string, Tracking> cycler;
+
+	joinpartspamsettings(int c, int s) : cycles(c), secs(s)
```
These parameters should be `unsigned int`.

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,231 @@
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		}
+
+		++cycler[mask].counter;
```
Each `cycler[mask]` causes a map lookup which is wasteful because you are only interacting with the same element throughout the method. To improve this, do something like:
`Tracking& tracking = cycler[mask];`
then replace `cycler[mask]` with `tracking` in the rest of the method.

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,231 @@
+ public:
+	unsigned int cycles;
+	unsigned int secs;
+	time_t lastcleanup;
```
This must be given a value before line 88 checks it. As I understand the initial value has to pass the check on that line for the code to work as intended.

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,231 @@
+				return true;
+			}
+		}
+		return false;
```
What you can do here to avoid repeated map lookups is to use the `find()` map method like this:

```c++
std::map<std::string, Tracking>::iterator it = cycler.find(mask);
if (it == cycler.end()) // 'mask' is not in the map
    return false;
Tracking& tracking = it->second;

// Rest of the code
if (ServerInstance->Time() > tracking.reset)
 ...
```

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,231 @@
+		if (ServerInstance->Time() - lastcleanup < 600)
+			return;
+
+		for (std::map<std::string, Tracking>::iterator i = cycler.begin(); i != cycler.end(); )
```
The same thing applies for `cycler[mask]` and `erase(key)` as above, they must find the element in the map but you already know where it is.
There's existing code showing how to do this more efficiently: https://github.com/inspircd/inspircd/blob/master/src/coremods/core_dns.cpp#L658-L665
That way, no extra strings get created and extra map lookups don't occur.

genius3000 commented:
@attilamolnar how's this looking now?

attilamolnar commented:
@genius3000 Code-wise looks ok. I see it is storing the ident along with the host. Do you intend it to work this way?

attilamolnar commented on 2.0/m_joinpartspam.cpp:
```diff
@@ -0,0 +1,243 @@
+				parameters.push_back(user->MakeWildHost());
+				ServerInstance->SendGlobalMode(parameters, ServerInstance->FakeClient);
+
+				user->WriteNumeric(474, "%s %s :Channel JoinPart Spam triggered (limit is %u cycles in %u secs)", user->nick.c_str(), chan->name.c_str(), jpss->cycles, jpss->secs);
```
Can you reword this to say "join/part spam" to look more natutal instead of "JoinPart Spam"?

genius3000 commented:
@attilamolnar yes, I intended the use of ident@host; Figured host may be too broad, and using nickname as well is too specific/bypassed.
I'll change the wording in the three Numerics, that does look better.
Good to squash the commits down to just one then as well?

attilamolnar commented:
Ok.

> Good to squash the commits down to just one then as well?

Yeah please do

genius3000 commented:
@atillamolnar Alright, squashed down to the one commit. Thank you very much for all the assistance.

attilamolnar commented:
Thanks for your contribution!
